croncat/src/streams/polling.rs

```diff
@@ -15,45 +18,62 @@ use crate::logging::info;
     duration: Duration,
-    block_stream_tx: BlockStreamTx,
-    mut shutdown_rx: ShutdownRx,
-    rpc_address: String,
-) {
-    info!("rpc_address {}", rpc_address);
-
+    block_stream_tx: &BlockStreamTx,
+    shutdown_rx: &ShutdownRx,
+    rpc_address: &String,
+) -> Result<(), Report> {
     let node_address: Url = rpc_address.parse().unwrap();
-    info!("node_address {}", node_address);
 
-    let rpc_client =
-        HttpClient::new(node_address).expect("Could not get http client for RPC node for polling");
+    info!("Polling connecting to {}", node_address);
 
-    let polling_loop_handle = tokio::task::spawn(async move {
-        loop {
-            let block_response = rpc_client
-                .latest_block()
-                .await
-                .expect("Could not fetch latest block");
-            let block_height = block_response.block.header.height.value();
-            info!("block_height {}", block_height);
+    let rpc_client = HttpClient::new(node_address.clone()).map_err(|err| {
+        eyre!(
+            "Could not get http client for RPC node for polling: {}",
+            err.detail()
+        )
+    })?;
+
+    info!("Polling connected to {}", node_address);
 
+    let block_stream_tx = block_stream_tx.clone();
+    let mut shutdown_rx = shutdown_rx.clone();
+
+    let polling_loop_handle: JoinHandle<Result<(), Report>> = tokio::task::spawn(async move {
+        loop {
+            let block = tokio::time::timeout(Duration::from_secs(15), rpc_client.latest_block())
+                .await?
+                .map_err(|err| eyre!("Timed out latest block: {}", err.detail()))?
+                .block;
+            info!(
+                "Polled block (height: {}) from {}",
+                block.header.height, block.header.time
+            );
             // Broadcast block height, will be received by …?
             // Currently getting:
             //   The application panicked (crashed).
             //   Message:  Failed to send block height from polling: SendError(..)
             // I think we need to have the block stream receiver (likely in )
-            block_stream_tx
-                .broadcast(block_response.block)
-                .await
-                .expect("Failed to send block height from polling");
-
+            block_stream_tx.broadcast(block).await?;
             // Wait
             sleep(duration).await;
         }
     });
 
-    // Allow this task to get shut down when a person types Ctrl+C
     tokio::select! {
-        _ = polling_loop_handle => {}
-        _ = shutdown_rx.recv() => {}
+        res = polling_loop_handle => {
+            match res? {
+                Ok(_) => {
+
+                    return Ok(())
+                }
+                Err(err) => {
+                    error!("Block polling loop failed: {}", err);
+                    return Err(err.into());
+                }
+            }
+        }
+        _ = shutdown_rx.recv() => {
+            return Ok(())
```

Comment:
```suggestion
            res?.map_err(|err| {
                error!("Block polling loop failed: {}", err);
                err
            })
        }
        _ = shutdown_rx.recv() => Ok(())
```

Reply:
So the reason I kept the full pattern and didn't use .map_err is because we might have work to do on shutdown/closure. In the ws stream module you can see the use case of that happening. For now it's a little unwieldy but I think I'd rather have the pattern be there in stone.

Reply:
Actually I think there is a way to simplify it!

Reply:
I see, thanks, thought they shutdown on `drop`